config/internal/configsource/configsource.go

```diff
@@ -0,0 +1,315 @@
+)
+
+// ApplyConfigSourcesParams holds the parameters for injecting data from the
+// given configuration sources into a configuration.
+type ApplyConfigSourcesParams struct {
+	// Set of configuration sources available to inject data into the configuration.
+	ConfigSources map[string]component.ConfigSource
+	// MaxRecursionDepth limits the maximum number of times that a configuration source
+	// can inject another into the config.
+	MaxRecursionDepth uint
+}
+
+// ApplyConfigSources takes a viper object with the configuration to which the
+// the configuration sources will be applied and the resulting configuration is
+// returned as a separated object.
+func ApplyConfigSources(ctx context.Context, v *viper.Viper, params ApplyConfigSourcesParams) (*viper.Viper, error) {
+
+	// Notify all sources that we are about to start.
+	for cfgSrcName, cfgSrc := range params.ConfigSources {
+		if err := cfgSrc.BeginSession(ctx); err != nil {
+			return nil, &cfgSrcError{
+				msg:  fmt.Sprintf("config source %s begin session error: %v", cfgSrcName, err),
+				code: errCfgSrcBeginSession,
+			}
+		}
+
+		// The scope of usage of the cfgSrc is the whole function so it is fine to defer
+		// the EndSession call to the function exit.
+		defer cfgSrc.EndSession(ctx)
+	}
+
+	srcCfg := v
+	var dstCfg *viper.Viper
+
+	var err error
+	done := false
+	for i := -1; i <= int(params.MaxRecursionDepth) && !done; i++ {
+		dstCfg = config.NewViper()
+		done, err = applyConfigSources(ctx, srcCfg, dstCfg, params.ConfigSources)
+		if err != nil {
+			return nil, err
+		}
+		srcCfg = dstCfg
+	}
+
+	if !done {
+		return nil, &cfgSrcError{
+			msg:  "config source recursion chain is too deep, couldn't fully expand the configuration",
+			code: errCfgSrcChainTooLong,
+		}
+	}
+
+	return dstCfg, nil
+}
+
+// cfgSrcError private error type used to accurate identify the type of error in tests.
+type cfgSrcError struct {
+	msg  string          // human readable error message.
+	code cfgSrcErrorCode // internal error code.
+}
+
+func (e *cfgSrcError) Error() string {
+	return e.msg
+}
+
+// These are errors that can be returned by ApplyConfigSources function.
+// Note that error codes are not part public API, they are for unit testing only.
+type cfgSrcErrorCode int
+
+const (
+	_ cfgSrcErrorCode = iota // skip 0, start errors codes from 1.
+	errCfgSrcChainTooLong
+	errOnlyMapAtRootLevel
+	errCfgSrcBeginSession
+	errCfgSrcNotFound
+	errCfgSrcApply
+)
+
+func applyConfigSources(ctx context.Context, srcCfg, dstCfg *viper.Viper, cfgSources map[string]component.ConfigSource) (bool, error) {
+	// Expand any item env vars in the config, do it every time so env vars
+	// added on previous pass are also handled.
+	expandEnvConfig(srcCfg)
+
+	done := true
+	appliedTags := make(map[string]struct{})
+
+	// It is possible to have config sources injection depending on other config sources:
+	//
+	// $lower_cfgsrc:
+	//   a: "just an example"
+	//   b:
+	//     $deeper_cfgsrc:
+	//       c: true
```

Comment:
So it is not possible to have this as value only as a key in the config? It does not seem to work if I have:

```
a: $cfgsrc
```

My understanding was the opposite and that we want only values to be cfgsources like envvar

Reply:
I will update this to go back to having the config sources as values - still pending.

Reply:
Resolving since this part was removed from the PR.